src/sentry/migrations/0869_backfill_open_periods.py

```diff
@@ -0,0 +1,205 @@
+    group_id: int,
+    status: int,
+    project_id: int,
+    first_seen: datetime,
+    activities: list[Any],
+    GroupOpenPeriod: Any,
+) -> list[Any]:
+    # Filter to REGRESSION and RESOLVED activties to find the bounds of each open period.
+    # The only UNRESOLVED activity we would care about is the first UNRESOLVED activity for the group creation,
+    # but we don't create an entry for that.
+    open_periods = []
+    start: datetime | None = None
+    end: datetime | None = None
+    end_activity: ActivityModelType | None = None
+
+    # Handle currently open period
+    if status == GroupStatus.UNRESOLVED and len(activities) > 0:
+        open_periods.append(
+            GroupOpenPeriod(
+                group_id=group_id,
+                project_id=project_id,
+                date_started=activities[0].datetime,
+                date_ended=None,
+                resolution_activity=None,
+                user_id=None,
+            )
+        )
+        activities = activities[1:]
+
+    for activity in activities:
+        if activity.type == ActivityType.SET_RESOLVED.value:
+            end = activity.datetime
+            end_activity = activity
+        elif activity.type == ActivityType.SET_REGRESSION.value:
+            start = activity.datetime
+            if end is None:
+                logger.error(
+                    "No end activity found for group open period backfill",
+                    extra={"group_id": group_id, "starting_activity": activity.id},
+                )
+            if start is not None and end is not None:
+                open_periods.append(
+                    GroupOpenPeriod(
+                        group_id=group_id,
+                        project_id=project_id,
+                        date_started=start,
+                        date_ended=end,
+                        resolution_activity=end_activity,
+                        user_id=end_activity.user_id if end_activity else None,
+                    )
+                )
+                end = None
+                end_activity = None
+    # Add the very first open period, which has no UNRESOLVED activity for the group creation
+    open_periods.append(
+        GroupOpenPeriod(
+            group_id=group_id,
+            project_id=project_id,
+            date_started=first_seen,
+            date_ended=end if end else None,
+            resolution_activity=end_activity,
+            user_id=end_activity.user_id if end_activity else None,
+        )
+    )
+    return open_periods
+
+
+def _backfill_group_open_periods(
+    apps: StateApps, group_data: list[tuple[int, datetime, int, int]]
+) -> None:
+    GroupOpenPeriod = apps.get_model("sentry", "GroupOpenPeriod")
+    Activity = apps.get_model("sentry", "Activity")
+
+    group_ids = [group_id for group_id, _, _, _ in group_data]
+    groups_with_open_periods = set(
+        GroupOpenPeriod.objects.filter(group_id__in=group_ids).values_list("group_id", flat=True)
```

Comment:
Could be worth making this query distinct so that we don't need to bring all the duplicate ids down